sky/serve/replica_managers.py

```diff
@@ -703,14 +758,20 @@ def _refresh_process_pool(self) -> None:
                         removal_reason = 'for preemption recovery'
                     else:
                         removal_reason = 'normally'
-                    logger.info(f'Replica {replica_id} removed from the '
-                                f'replica table {removal_reason}.')
+                # Don't keep failed record for version mismatch replicas,
+                # since user should fixed the error before update.
+                elif info.version != self.latest_version:
+                    serve_state.remove_replica(self._service_name, replica_id)
+                    removal_reason = 'for old version'
                 else:
                     logger.info(f'Termination of replica {replica_id} '
                                 'finished. Replica info is kept since some '
                                 'failure detected.')
                     serve_state.add_or_update_replica(self._service_name,
                                                       replica_id, info)
+                if removal_reason is not None:
+                    logger.info(f'Replica {replica_id} removed from the '
```

Comment:
We should move the `serve_state.remove_replica(self._service_name, replica_id)` in L756 to here too.